### PR TITLE
Minimize image and 1 step closer for release to Github pages

### LINKED ISSUE #5: Governor is set to ondemand instead of performance out of the box
As I'm testing the image for #4, I see that the CPU governor is set to ondemand after I flashed the image. This shouldn't be the case, as `setup/phase1.sh` disables the ondemand.service startup file and https://github.com/ros-realtime/linux-real-time-kernel-builder/commit/483090dfa69de42fc63a5474da447c709d73f669 adds performance to always be enabled? 

Not quite sure what's going on. Will have to double check later. Not the highest priority item right now.

```
ubuntu@ubuntu:~$ uname -a
Linux ubuntu 5.4.140-rt64 #1 SMP PREEMPT_RT Sat Oct 16 17:57:52 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@ubuntu:~$ sudo systemctl status ondemand
● ondemand.service - Set the CPU Frequency Scaling governor
     Loaded: loaded (/lib/systemd/system/ondemand.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
ubuntu@ubuntu:~$ cat /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
ondemand
ondemand
ondemand
ondemand
```

### PR DESCRIPTION
**Pull request by shuhaowu**
As discussed in the last RTWG meeting, we should get an image hosted on Github by stripping it down, so I've done this:

1. Removed Linux kernel debug package from being installed, which saves 3GB from the raw image.
2. Do not install ROS galactic by default, but the ROS apt repo remains to be setup.
3. Removed another few packages to save space.
4. Removed stock linux kernel (saved a few hundred MB)
5. Fixed a bunch of file permissions from the `rootfs` directory. Previously it was owned by uid=host_owner and gid=hostt_group instead of root:root.
6. Moved `builder/cleanup.sh` to `scripts/cleanup.sh` and updated readme slightly. This script is used to cleanup your system if the build fails and you want to reset instead of resume.

I've also noted that `pigz` can be used to gzip the final image in parallel, resulting in an image size of 1.1GB, which I think is good enough for Github releases. Final image size on my local machine is 1,232,325,713 bytes.

```
-rw-r--r-- 1 root root 1232325713 Feb  6 20:13 ubuntu-20.04.3-rt-ros2-galactic-arm64+raspi.img.gz
```

Cyclictest results (after fixing #5 manually):

```
ubuntu@ubuntu:~$ uname -a
Linux ubuntu 5.4.140-rt64 #1 SMP PREEMPT_RT Sat Oct 16 17:57:52 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@ubuntu:~$ sudo vcgencmd measure_clock arm
frequency(48)=1500345728
ubuntu@ubuntu:~$ cat /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
performance
performance
performance
performance
ubuntu@ubuntu:~$ sudo cyclictest --mlockall --smp --priority=80 --interval=200 --distance=0 -D 60s
# /dev/cpu_dma_latency set to 0us
policy: fifo: loadavg: 0.76 0.28 0.43 2/168 32490          

T: 0 (32484) P:80 I:200 C: 299991 Min:      5 Act:    8 Avg:    8 Max:      60
T: 1 (32485) P:80 I:200 C: 299934 Min:      6 Act:    8 Avg:    8 Max:      63
T: 2 (32486) P:80 I:200 C: 299875 Min:      6 Act:    8 Avg:    8 Max:      45
T: 3 (32487) P:80 I:200 C: 299818 Min:      6 Act:    9 Avg:    8 Max:      35
```

What's next after this:

- Fix #5.
- Setup /etc/security/limits.conf. Presumably we want the `ubuntu` user to have permissions to run RT processes.
- Fixed the upstream kernel hacks (vmlinuz and the LINUX_RT_VERSION_ACTUALLY hacks)
- Other stuff listed in #1 
- Tag and release an image to GH releases

### CODE REVIEW

File: focal-rt-ros2/phase1-outside.sh [hunk 1]
```diff
@@ -12,11 +12,14 @@ fi
 
 unzip $filename
 
+# Do not install the debug symbols as this causes an additional 3GB of storage
+# needed, which is will cause us to go over the 2GB limit for Github releases.
+rm linux*dbg*.deb
```
Comment (LanderU) [inline focal-rt-ros2/phase1-outside.sh:17]: Nice!

File: focal-rt-ros2/phase1-outside.sh [hunk 2]
```diff
@@ -12,11 +12,14 @@ fi
...
+ls -l
+
 # These deb files will be installed by the phase1 script.
 # The /setup folder in the chroot will be deleted by phase2 script
 cp linux-*.deb $CHROOT_PATH/setup/
 rm linux-*.deb
 
-echo "PINNED_CPU_FREQUENCY=${PINNED_CPU_FREQUENCY}" > $CHROOT_PATH/etc/default/cpu-frequency
```
Comment (LanderU) [inline focal-rt-ros2/phase1-outside.sh:20]: Why this change?

File: focal-rt-ros2/rootfs/setup/phase1.sh [hunk 3]
```diff
@@ -6,13 +6,39 @@ set -e -o pipefail
...
+# Remove some packages that are likely not needed:
+# - snapd: no one packages their robot apps with snap, right?
+# - fwupd: I don't think we need to update devices firmware like a logitech mouse, and it also uses like 20MB of RAM...
+# - cryptsetup: don't need to setup disk encryption. Also, causes build failures on some host system configurations.
+# - mdadm: don't need to setup raid.
+# - Stock linux kernel: for obvious reasons
+apt-get purge --autoremove -y \
+  cryptsetup \
+  fwupd \
+  linux-headers-raspi \
```
Comment (LanderU) [inline focal-rt-ros2/rootfs/setup/phase1.sh:22]: I was about to suggest also to remove the old kernel/modules.

File: focal-rt-ros2/rootfs/setup/phase1.sh [hunk 4]
```diff
@@ -6,13 +6,39 @@ set -e -o pipefail
...
+  xfsprogs
+
 # Setting up PREEMPT_RT kernel
 cd /setup
 sudo dpkg -i linux-*.deb
 
 ln -s -f /boot/vmlinuz-${LINUX_RT_VERSION_ACTUALLY} /boot/vmlinuz
 ln -s -f /boot/initrd.img-${LINUX_RT_VERSION_ACTUALLY} /boot/initrd.img
 
+# TODO: This should be removable following https://github.com/ros-realtime/linux-real-time-kernel-builder/pull/32
```
Comment (LanderU) [inline focal-rt-ros2/rootfs/setup/phase1.sh:41]: We can merge that PR and remove these extra steps.

File: focal-rt-ros2/rootfs/setup/phase1.sh [hunk 5]
```diff
@@ -24,18 +50,13 @@ systemctl enable cpu-frequency
...
-# - fwupd: I don't think we need to update devices firmware like a logitech mouse, and it also uses like 20MB of RAM...
-apt-get purge --autoremove -y snapd fwupd
-
-# Install ROS2
+# Setup ROS distro without installing ROS to save space on the image.
 sudo curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key  -o /usr/share/keyrings/ros-archive-keyring.gpg
-
-sudo apt-get update
-sudo apt-get install -y ros-galactic-ros-base
+# sudo apt-get install -y ros-galactic-ros-base
```
Comment (LanderU) [inline focal-rt-ros2/rootfs/setup/phase1.sh:55]:
```suggestion
sudo apt-get install -y ros-galactic-ros-base
```

With ROS 2 installed, do we increase the image size a lot?

File: focal-rt-ros2/vars.sh [hunk 6]
```diff
@@ -40,10 +41,10 @@ IMAGE_PARTITION_MOUNTS=(
 
 # This is passed to truncate --size=$IMAGE_SIZE when operating against the .img file.
 # TODO: Note that always the last partition will be expanded as I just call truncate for now.
-IMAGE_SIZE=8G
+IMAGE_SIZE=4G
 
 # Absolute path of the output image on the host.
-OUTPUT_FILENAME=$(pwd)/out/ubuntu-20.04.3-rt-ros2-galactic-arm64+raspi.img
+OUTPUT_FILENAME=$(pwd)/out/ubuntu-20.04.3-rt-ros-arm64+raspi.img
```
Comment (LanderU) [inline focal-rt-ros2/vars.sh:47]:
```suggestion
OUTPUT_FILENAME=$(pwd)/out/ubuntu-20.04.3-rt-ros2-arm64+raspi.img
```

File: focal-rt-ros2/phase1-outside.sh (hunk 2 above)
Comment (shuhaowu) [reply focal-rt-ros2/phase1-outside.sh:20]: This script is run on the host instead of in the "container" that is the rootfs of the system. This means the permission of this file end up inheriting whatever user executed this, which is a minor cosmetic problem. I moved this line of code to run inside the container, in `phase1.sh`.

Sorry I didn't elaborate about this as I'm still sorting out a few small things + testing and didn't have time to write a proper description.

File: focal-rt-ros2/rootfs/setup/phase1.sh (hunk 5 above)
Comment (shuhaowu) [reply focal-rt-ros2/rootfs/setup/phase1.sh:55]:
```
-rw-r--r-- 1 root root  1.4G Feb  6 15:56 ubuntu-20.04.3-rt-ros2-galactic-arm64+raspi.img.gz
-rw-r--r-- 1 root root 1015M Feb  6 02:02 ubuntu-20.04.3-rt-ros-arm64+raspi.img.gz
```

~400MB after compression. 1.4GB should still be good, as GH releases limit is [2GB](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases#storage-and-bandwidth-quotas). I'll reinstate this.

**Comment by shuhaowu** [comment]
I think this is about as small as it can be. Looking at the largest packages remaining on the system after installing ros-galactic:

```
13184   binutils-aarch64-linux-gnu
14680   systemd
15761   libc6-dev
16138   cmake
17196   locales
17226   perl-modules-5.30
17841   libpython3.8-dev
19122   cpp-9
19553   iso-codes
19769   python3-numpy
22523   gcc-9
23751   linux-firmware-raspi2
26547   libperl5.30
30765   vim-runtime
32584   libicu66
34892   git
74898   linux-headers-5.4.140-rt64
196837  linux-image-5.4.140-rt64
574082  linux-firmware
```

linux-firmware is >500MB, but there are files in there that are required if you plug in different USB devices, so I don't think removing it is a good idea. Doesn't matter that much tho, my final run with galactic + only linux rt gave an image that's only 1.2GB, well within the applicable range for Github releases:

```
-rw-r--r-- 1 root root 1232325713 Feb  6 20:13 ubuntu-20.04.3-rt-ros2-galactic-arm64+raspi.img.gz
```

I need to check to ensure this final image works on my raspberry pi before marking this PR as ready to be reviewed.

File: focal-rt-ros2/rootfs/setup/phase1.sh (hunk 5 above)
Comment (LanderU) [reply focal-rt-ros2/rootfs/setup/phase1.sh]: Thank you!

File: focal-rt-ros2/vars.sh (hunk 6 above)
Comment (LanderU) [reply focal-rt-ros2/vars.sh]: Thank you!

**Comment by LanderU** [comment]
@shuhaowu, this looks really promising. We should sync because I was doing more or less the same changes. We're going to validate this implementation and we can merge it! Thanks a lot!

File: focal-rt-ros2/rootfs/setup/phase1.sh (hunk 3 above)
Comment (shuhaowu) [reply focal-rt-ros2/rootfs/setup/phase1.sh:22]: Done below.

File: focal-rt-ros2/rootfs/setup/phase1.sh (hunk 4 above)
Comment (shuhaowu) [reply focal-rt-ros2/rootfs/setup/phase1.sh:41]: Let's merge this one and then the kernel changes? We need to do another kernel release. Right now the release version tagged on github doesn't match the file names built, which requires another hack in here (check the variable `LINUX_RT_VERSION_ACTUALLY`. Once that's all done, we can come back and fix these.

**Comment by shuhaowu** [comment]
> this looks really promising. We should sync because I was doing more or less the same changes. We're going to validate this implementation and we can merge it! Thanks a lot!

I see the next RTWG meeting is next week. We can sync up then?

I've cleaned up the git commit history here by squashing stuff and also updated the PR description with everything I've done, along with what's next. Github CI on my repo also [passed this branch](https://github.com/shuhaowu/ros-realtime-rpi4-image/actions).  If I can get a review I can merge this and we can move forward perhaps with the RT kernel bug fixes.

**Review by carlossvg — APPROVED** [review]
LGTM
